File: backend/entity/report_digest.go
package entity

import (
	"time"

	"gorm.io/gorm"
)

// AppReportDigestSchedule is a recurring "auto report": on its schedule the
// backend generates the executive PDF report (the same one the Report page
// produces) and delivers it to the chosen channel — email or LINE. It reuses
// the manual send-to-email / send-to-LINE pipeline, just fired by a scheduler
// instead of a button.
type AppReportDigestSchedule struct {
	ID   uint   `gorm:"primaryKey;autoIncrement" json:"id"`
	Name string `gorm:"column:name;not null" json:"name"`
	// Channel is how the report is delivered: "email" | "line".
	Channel string `gorm:"column:channel;not null" json:"channel"`
	// Frequency: "weekly" | "monthly" | "yearly".
	Frequency string `gorm:"column:frequency;not null" json:"frequency"`
	Hour      int    `gorm:"column:hour;not null;default:8" json:"hour"`   // 0-23
	Minute    int    `gorm:"column:minute;not null;default:0" json:"minute"` // 0-59
	// DayOfWeek is 0=Sunday .. 6=Saturday, used when Frequency=weekly.
	DayOfWeek int `gorm:"column:day_of_week;not null;default:1" json:"day_of_week"`
	// DayOfMonth is 1-31, used when Frequency=monthly (clamped to the month's length).
	DayOfMonth int `gorm:"column:day_of_month;not null;default:1" json:"day_of_month"`
	// Month (1-12) + Day (1-31) are used when Frequency=yearly.
	Month int `gorm:"column:month;not null;default:1" json:"month"`
	Day   int `gorm:"column:day;not null;default:1" json:"day"`
	// Timezone is an IANA name (e.g. "Asia/Bangkok"); NextRunAt is computed in it.
	Timezone string `gorm:"column:timezone;not null;default:'Asia/Bangkok'" json:"timezone"`
	// EmailTo is a comma-separated list of recipient emails (Channel=email).
	EmailTo string `gorm:"column:email_to;type:text" json:"email_to"`
	// LineNotificationIDs is a comma-separated list of app_notification ids the
	// report link is pushed to (Channel=line).
	LineNotificationIDs string `gorm:"column:line_notification_ids;type:text" json:"line_notification_ids"`

	Enabled    bool       `gorm:"column:enabled;default:true" json:"enabled"`
	LastRunAt  *time.Time `gorm:"column:last_run_at" json:"last_run_at"`
	LastStatus string     `gorm:"column:last_status" json:"last_status"` // "", "ok", "failed"
	LastError  string     `gorm:"column:last_error;type:text" json:"last_error"`
	NextRunAt  *time.Time `gorm:"column:next_run_at" json:"next_run_at"`

	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

File: frontend/src/services/reportdigest.ts
import { baseApi } from "./api";

// Auto-Report Digest: scheduled generation of the executive PDF report,
// delivered to email or LINE on a weekly / monthly / yearly cadence.

export type DigestChannel = "email" | "line";
export type DigestFrequency = "weekly" | "monthly" | "yearly";

export type ReportDigestDTO = {
  id: number;
  name: string;
  channel: DigestChannel;
  frequency: DigestFrequency;
  hour: number;
  minute: number;
  day_of_week: number; // 0=Sun..6=Sat (weekly)
  day_of_month: number; // 1-31 (monthly)
  month: number; // 1-12 (yearly)
  day: number; // 1-31 (yearly)
  timezone: string;
  email_to: string; // comma-separated
  line_notification_ids: string; // comma-separated app_notification ids
  enabled: boolean;
  last_run_at: string | null;
  last_status: string; // "", "ok", "failed"
  last_error: string;
  next_run_at: string | null;
  created_at: string;
  updated_at: string;
};

export type ReportDigestInput = {
  name: string;
  channel: DigestChannel;
  frequency: DigestFrequency;
  hour: number;
  minute: number;
  day_of_week: number;
  day_of_month: number;
  month: number;
  day: number;
  timezone: string;
  email_to: string;
  line_notification_ids: string;
  enabled: boolean;
};

export const ListReportDigests = async (): Promise<ReportDigestDTO[]> => {
  try {
    const res = await baseApi.get("/report-digests");
    const data = res.data?.data ?? [];
    return Array.isArray(data) ? (data as ReportDigestDTO[]) : [];
  } catch (e) {
    console.error("ListReportDigests error:", e);
    return [];
  }
};

// Create/Update/Delete/Run are left to throw so callers surface the backend error.
export const CreateReportDigest = async (input: ReportDigestInput): Promise<ReportDigestDTO> => {
  const res = await baseApi.post("/report-digests", input);
  return (res.data?.data ?? res.data) as ReportDigestDTO;
};

export const UpdateReportDigest = async (id: number, input: ReportDigestInput): Promise<ReportDigestDTO> => {
  const res = await baseApi.patch(`/report-digests/${id}`, input);
  return (res.data?.data ?? res.data) as ReportDigestDTO;
};

export const DeleteReportDigest = async (id: number): Promise<void> => {
  await baseApi.delete(`/report-digests/${id}`);
};

// Send this digest right now (synchronous — resolves after it's actually sent).
export const RunReportDigestNow = async (id: number): Promise<void> => {
  await baseApi.post(`/report-digests/${id}/run`);
};
